exec/cnex/lib/mmap_posix.c

```diff
@@ -0,0 +1,184 @@
+#include "mmap.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "cell.h"
+#include "exec.h"
+#include "nstring.h"
+#include "number.h"
+#include "object.h"
+#include "stack.h"
+
+
+typedef struct tagTMmapObject {
+    int fd;
+    size_t len;
+    char *view;
+} TMmapObject;
+
+
+static TMmapObject *check_file(TExecutor *exec, Object *pp)
+{
+    if (pp == NULL || pp->ptr == NULL || ((TMmapObject*)(pp->ptr))->fd < 0) {
+        exec->rtl_raise(exec, "MmapException.InvalidFile", "", BID_ZERO);
+        return NULL;
+    }
+    return pp->ptr;
+}
+
+void object_releaseMMapObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                free(o->ptr);
```

Comment:
I can't help but think that there's something missing here. How does the file get closed if all the references to the object go away?

Reply:
Fixed.  If `(o->ptr)->fd > 0` it will unmap the view, and close the file.